hack/gotagsrewrite/main.go

```diff
@@ -0,0 +1,312 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"fmt"
+	"go/ast"
+	"go/format"
+	"go/token"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/fatih/structtag"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Fprintln(os.Stderr, "usage: gotagsrewrite <path>")
+		os.Exit(1)
+	}
+
+	path := os.Args[1]
+	if path == "" {
+		fmt.Fprintln(os.Stderr, "usage: gotagsrewrite <path>")
+		os.Exit(1)
+	}
+
+	if err := Run(path); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+}
+
+// Run runs the main logic of the program.
+func Run(path string) error {
+	paths, err := findGoFiles(path)
+	if err != nil {
+		return fmt.Errorf("failed to find Go files: %w", err)
+	}
+
+	filesAndStructs := map[string]fileStruct{}
+	tokenSet := token.NewFileSet()
+
+	for _, path := range paths {
+		parsedAST, err := parseAst(tokenSet, path) //nolint:govet
+		if err != nil {
+			return fmt.Errorf("failed to parse AST for %s: %w", path, err)
+		}
+
+		structs := filterStructWithFields(parsedAST, func(n *ast.TypeSpec, t *ast.StructType) bool {
+			return strings.HasSuffix(n.Name.Name, "SpecSpec") && isCapitalCase(n.Name.Name)
```

Comment:
I feel this doesn't cover _all_ of the structs we need to handle. Plus there might be embedded structs into the ones we analyze?

we can also do some kind of docblock tag to tag the specs, and anything which is embedded should be also covered?

Reply:
we could also look for whatever uses `typed.Resource[<Spec>]`

Reply:
> we can also do some kind of docblock tag to tag the specs

I think it's a best approach, since its easy enough and quite flexible. It requires _some_ manual work, but nothing particularly hard.

Reply:
> we could also look for whatever uses typed.Resource[<Spec>]

i think we then need to slowly update all controllers to use this, since the old ones doesn't used the typed.Resource